Add dicom concatenations

jcupitt opened the pull request:
Layers in large DICOM WSI files are commonly split into separate files to make transmission easier. This PR adds support for `TILED_FULL` and `TILED_SPARSE` scans with split levels. 

Referring issue:

See https://github.com/openslide/openslide/issues/511

Overview:

1. The openslide DICOM loader has separate structures for `dicom_level` and `dicom_file`, but only allows one `dicom_file` per level. This PR adds an array of `dicom_file` for each level, with an array `int file_index[n_tiles]` giving the index of the file containing each tile.
2. During directory scan, files are attached to their matching level.
3. In `read_tile()`, the `file_index` array is constructed on first use (protected by a mutex) using a new libdicom v1.2.0 function `dcm_filehandle_get_frame_number()` which allows fast readout of the internal libdicom tile index. Delaying index construction until use gives a very nice speedup for slide open. 
4. Small changes around testing the various UIDs which link files and levels together.

The new libdicom has two relevant changes:

1. The new `dcm_filehandle_get_frame_number()` as noted above. 
2. Support for the `ConcatenationFrameOffsetNumber` field, for `TILED_FULL` catenations.

Sample data:

I have some cc0 sample files kindly donated by Geneva University Hospital, and I think dclunie has many others. I'll upload one of them in a moment.

Geneva have been using this branch for a few weeks and say it's working well for them.

TODO:

- [x] update openslide to require libdicom 1.2.0, update build files, update wrap
- [x] fix failures in current tests
- [ ] update documentation
- [x] add new tests and sample data (maybe this should be a separate PR?)
- [ ] squash, sign and split PR (I'll do this at the end)

jcupitt committed:
Add DICOM catenation support

Add support for DICOM files where levels have been split into many
files.

Signed-off-by: John Cupitt <jcupitt@gmail.com>
Signed-off-by: Benjamin Gilbert <bgilbert@cs.cmu.edu>

jcupitt commented:
Here's a sample file:

http://www.rollthepotato.net/~john/DS_A07R_01S.zip

5gb, level 4 is split into six parts, level 5 is split into two.

bgilbert commented:
Thanks for the PR!  Some preliminary things:

- By policy, I don't look at diffs until they're signed off.  I know you're good for it, but if the signoffs don't specifically need to be deferred, ideally you could rebase to drop the merge commits and sign off.
- We should probably have a separate preliminary PR to update the libdicom wrap and fix tests, due to ImagingDataCommons/libdicom@7381e3ea1907780458376162b41922260d82107a.  Automation would normally handle this, but it won't fix the tests of course.  Tests should use regexes to support both old and new libdicom.
-  Could you submit the sample data through the [official form](https://openslide.org/submit/) so we have a record of the license and details?

> Delaying index construction until use gives a very nice speedup for slide open.

Without having seen the code yet, I'll note that I'm skeptical.  We do have one case in the existing code (Hamamatsu VMS restart marker scan) which is deferred from open time because it's too I/O-intensive to do up front, but in that case the code runs incrementally in a background thread, and has a debug flag to force it to run eagerly.  It sounds as though this PR does the full work on the first `read_region`?  I'd rather not introduce some surprising tail latency.

> add new tests and sample data (maybe this should be a separate PR?)

Actually, adding in this PR is probably better, once the data goes into openslide-testdata.

jcupitt commented:
Here's another test file (15gb!!):

ftp://medical.nema.org/MEDICAL/Dicom/DataSets/WG26/3DHISTECH/FL_20191127/DICOM_TILED_SPARSE/MouseKidney_F24630_full

It's part of the 2023 connectathon, so it's recent, from the scanner manufacturer, and CC0. It doesn't work and probably should, I'll investigate.

jcupitt commented:
Thanks for the feedback!

I'll try to get libdicom 1.2.0 released in the next couple of days, then make a preliminary PR on openslide to use the new version.

I've chased the donor of the sample data to try to get the upload done.

I've removed the lazy index build, rebased and (this time) remembered to sign off the commit.

jcupitt commented:
For reference, I timed lazy index vs. index on open.

Right now (index on open) with a 6GB test scan (3Dhistech, levels split to stay within 500mb per file, no offset table), I see:

```
$ time openslide-show-properties 4_2 > /dev/null

real	0m4.527s
user	0m3.222s
sys	0m1.304s
```

If I put the lazy index back I see:

```
$ time openslide-show-properties 4_2 > /dev/null

real	0m0.016s
user	0m0.007s
sys	0m0.009s
```

And fetching a tile from level 4 with index on open:

```
$ time openslide-write-png 4_2 100 100 4 512 512 x.png

real	0m4.583s
user	0m3.333s
sys	0m1.249s
```

And with lazy index:

```
$ time openslide-write-png 4_2 100 100 4 512 512 x.png

real	0m0.048s
user	0m0.033s
sys	0m0.015s
```

Being able to get a low-ish resolution image on the screen quickly is really nice for interactive use.

jcupitt commented:
I've released libdicom v1.2.0 and made a preliminary PR to update openslide:

https://github.com/openslide/openslide/pull/658

Once that's merged I'll rebase this PR and remove the "draft" tag.

bgilbert committed:
dicom: drop unused level lock

Signed-off-by: Benjamin Gilbert <bgilbert@cs.cmu.edu>

jcupitt committed:
add samplesperpixel failure

we were not failing for incorrect samplesperpixel

Signed-off-by: John Cupitt <jcupitt@gmail.com>

jcupitt committed:
fix error messages in tests

Signed-off-by: John Cupitt <jcupitt@gmail.com>

jcupitt committed:
ignore levels with a duplicate SOPInstanceUID

since they are probably just duplicated files

Signed-off-by: John Cupitt <jcupitt@gmail.com>

jcupitt commented:
All tests are now passing, I think.

We could add some more tests I guess, but maybe that should be in a followup PR?

I've not done a final squash yet. This seems like a fairly simple PR to me, so I'd probably make it a single commit. Is that OK?

jcupitt committed:
test SOPInstanceUIDOfConcatenationSource

For catenated images, we need to test
SOPInstanceUIDOfConcatenationSource, not SOPInstanceUID, to check that
files are in the same level.

Signed-off-by: John Cupitt <jcupitt@gmail.com>

jcupitt committed:
remove two trailing spaces

Signed-off-by: John Cupitt <jcupitt@gmail.com>

choosehappy commented:
Would it be possible to bump the merge of this PR? 

we've been compiling and using it locally - but in our evolving CI/CD pipelines that local compilation is adding some complexity that would be removed post release of an official version  

we need this updated functionality for our clinical routine workflow. I can say that we've been using it internally for at least 4 months on ~1.5k slides per day without any issues - so functionality testing wise, seems rock solid!

Thanks

bgilbert commented:
Thanks for the experience report!  There's some infrastructure work I'm trying to finish before landing any more test cases, so I'm afraid it may be a bit, but I haven't lost track of this.  For your CI pipeline, could you build the necessary binaries once and stash them somewhere for reuse?

choosehappy commented:
great, thanks for the update.  happy to provide any more information from our side if useful

indeed - compiling and stashing is emerging as our to go method : ) this does create some additional complexities with versioning and share our tools externally with other hospitals, but we're managing. 

naturally a bit more elegant to have a public repo version we can just pull  and use directly ;-)

SebastianRiechert2 commented:
+1 on the request to merge this. it would reduce the complexity of our docker builds, since this feature is the last dependency holdout we can't just install from conda/pypi

bgilbert commented:
Rebased.  I'll go through this and add some cleanups.

bgilbert committed:
dicom: textual cleanups

Signed-off-by: Benjamin Gilbert <bgilbert@cs.cmu.edu>

bgilbert committed:
dicom: fix first-tile assumptions in paint_region

paint_region() was returning immediately if the first tile in the region
was sparse, and otherwise was only suspending I/O to that one tile's file
after painting.  Properly wrap I/O in dicom_file_io, keep track of all
the ones used during paint, and put them on the way out.

Signed-off-by: Benjamin Gilbert <bgilbert@cs.cmu.edu>

bgilbert committed:
dicom: report missing tiles in multi-file levels to the grid

Signed-off-by: Benjamin Gilbert <bgilbert@cs.cmu.edu>

bgilbert committed:
dicom: use 16 bits for file number

InConcatenationNumber has a VR of US and zero is forbidden, so there can't
be more than 65535 files.

Signed-off-by: Benjamin Gilbert <bgilbert@cs.cmu.edu>

bgilbert committed:
dicom: don't use GPtrArray for level files array at runtime

Follow the typical pattern of converting GPtrArray to a real array late
in open.

Signed-off-by: Benjamin Gilbert <bgilbert@cs.cmu.edu>

bgilbert committed:
dicom: clean up UID tests

Restore check for duplicate SOP Instance UIDs in a level.  Require
associated images not to be in a concatenation.  Switch level membership
check from SOPInstanceUIDOfConcatenationSource to ConcatenationUID since
we want to ensure all SOP Instances in a level are really from the same
concatenation.

Signed-off-by: Benjamin Gilbert <bgilbert@cs.cmu.edu>

bgilbert committed:
dicom: check level file count

Require that we have all files in a concatenation, if the concatenation
tells us the quantity.  Require that single-file levels are not in a
concatenation.

Signed-off-by: Benjamin Gilbert <bgilbert@cs.cmu.edu>

bgilbert committed:
dicom: order concatenation files with InConcatenationNumber

This ensures we always read slide properties from the same file,
independent of readdir order, if level 0 is a concatenation (which is
likely if any concatenations are present).  It also allows us to check for
missing files (other than trailing ones) if InConcatenationTotalNumber is
omitted.

Signed-off-by: Benjamin Gilbert <bgilbert@cs.cmu.edu>

bgilbert committed:
dicom: check for unexpected libdicom errors when building tile index

Signed-off-by: Benjamin Gilbert <bgilbert@cs.cmu.edu>

bgilbert committed:
dicom: suspend I/O after building tile index

Suspend when we're done with a file, rather than carrying conditional logic
to do it as soon as we can.

Signed-off-by: Benjamin Gilbert <bgilbert@cs.cmu.edu>

bgilbert committed:
dicom: fix error returns

Signed-off-by: Benjamin Gilbert <bgilbert@cs.cmu.edu>

bgilbert committed:
dicom: add concatenation tests

Signed-off-by: Benjamin Gilbert <bgilbert@cs.cmu.edu>

bgilbert committed:
dicom: undo allowing SamplesPerPixel == 1

Sup. 145 only allows a SamplesPerPixel of 1 if PhotometricInterpretation
is MONOCHROME2, and we don't permit that, so I don't think the looser
validation adds support for any slides.

Signed-off-by: Benjamin Gilbert <bgilbert@cs.cmu.edu>

bgilbert commented on src/openslide-vendor-dicom.c:
```diff
@@ -829,13 +902,32 @@ static bool maybe_add_file(openslide_t *osr,
+      return false;
+  }
+  switch (value) {
+  case 1:
```
Was this change meant to be included in this PR?  If `SamplesPerPixel` is 1, Sup. 145 requires `PhotometricInterpretation` to be `MONOCHROME2`.  Since OpenSlide doesn't allow that, I don't think allowing `SamplesPerPixel == 1` here adds support for any additional slides.

bgilbert replied:
I've removed this part for now.

bgilbert committed:
dicom: finalize levels earlier

Signed-off-by: Benjamin Gilbert <bgilbert@cs.cmu.edu>

bgilbert committed:
dicom: avoid g_autoptr() in level struct creation

Level creation is infallible and this cleans things up a bit.

Signed-off-by: Benjamin Gilbert <bgilbert@cs.cmu.edu>

bgilbert commented:
The changes were mostly refactoring and stricter metadata validation.  There was one substantive bug on well-formed datasets: if the `x` and `y` arguments to `openslide_read_region()` pointed to a sparse (missing) tile of a concatenated level, the call would not return any pixel data for the whole region.

bgilbert commented:
Thanks for implementing this!
